File: sdk/php/src/Reports.php
<?php
declare(strict_types=1);
namespace MerchantPlatform;

final class Reports {
    public static function verify(string $raw,array $report,array $publicKeys): void{$digest=hash('sha256',$raw,true);if(($report['status']??'')!=='ready'||!hash_equals((string)($report['object_sha256']??''),bin2hex($digest)))throw new \RuntimeException('reconciliation report digest mismatch');if(isset($report['object_size_bytes'])&&(int)$report['object_size_bytes']!==strlen($raw))throw new \RuntimeException('reconciliation report size mismatch');$keyId=(string)($report['signing_key_id']??'');if(!isset($publicKeys[$keyId]))throw new \RuntimeException('unknown reconciliation signing key: '.$keyId);$signature=self::base64UrlDecode((string)$report['signature']);$message="merchant-reconciliation-jsonl-v1\0".$report['id']."\0".$report['snapshot_ledger_sequence']."\0".$digest;if(!function_exists('sodium_crypto_sign_verify_detached'))throw new \RuntimeException('ext-sodium is required for report verification');if(!sodium_crypto_sign_verify_detached($signature,$message,$publicKeys[$keyId]))throw new \RuntimeException('reconciliation report signature mismatch');}
    private static function base64UrlDecode(string $value): string{$decoded=base64_decode(strtr($value,'-_','+/').str_repeat('=',(4-strlen($value)%4)%4),true);if($decoded===false)throw new \RuntimeException('invalid report signature');return $decoded;}
}

File: sdk/python/src/merchant_platform/reports.py
import base64
import hashlib
from typing import Mapping, Any

def reconciliation_signature_message(report_id: str, snapshot_ledger_sequence: str, digest: bytes) -> bytes:
    if len(digest) != 32: raise ValueError("SHA-256 digest must be 32 bytes")
    return b"merchant-reconciliation-jsonl-v1\0" + report_id.encode() + b"\0" + snapshot_ledger_sequence.encode() + b"\0" + digest

def verify_reconciliation_report(raw: bytes, report: Mapping[str, Any], public_keys: Mapping[str, bytes]) -> None:
    if report.get("status") != "ready": raise ValueError("report is not ready")
    digest = hashlib.sha256(raw).digest()
    if digest.hex() != report.get("object_sha256"): raise ValueError("reconciliation report digest mismatch")
    if "object_size_bytes" in report and int(report["object_size_bytes"]) != len(raw): raise ValueError("reconciliation report size mismatch")
    key_id = report.get("signing_key_id"); public_key = public_keys.get(key_id)
    if public_key is None: raise ValueError("unknown reconciliation signing key: " + str(key_id))
    signature = base64.urlsafe_b64decode(str(report["signature"]) + "=" * (-len(str(report["signature"])) % 4))
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
    except ImportError as error:
        raise RuntimeError("install merchant-platform-sdk[reports] for Ed25519 verification") from error
    Ed25519PublicKey.from_public_bytes(public_key).verify(signature, reconciliation_signature_message(str(report["id"]), str(report["snapshot_ledger_sequence"]), digest))
